server/src/main/java/org/elasticsearch/index/mapper/AnnotatedTextFieldMapper.java

```diff
@@ -0,0 +1,815 @@
+            currentAnnotationIndex++;
+            if (currentAnnotationIndex < annotatedText.numAnnotations()) {
+                nextAnnotationForInjection = annotatedText.getAnnotation(currentAnnotationIndex);
+            } else {
+                nextAnnotationForInjection = null;
+            }
+        }
+
+      }
+  
+
+    public static final class AnnotatedTextFieldType extends StringFieldType {
+
+        private boolean fielddata;
+        private double fielddataMinFrequency;
+        private double fielddataMaxFrequency;
+        private int fielddataMinSegmentSize;
+        private TextFieldMapper.PrefixFieldType prefixFieldType;
+
+        public AnnotatedTextFieldType() {
+            setTokenized(true);
+            fielddata = false;
+            fielddataMinFrequency = Defaults.FIELDDATA_MIN_FREQUENCY;
+            fielddataMaxFrequency = Defaults.FIELDDATA_MAX_FREQUENCY;
+            fielddataMinSegmentSize = Defaults.FIELDDATA_MIN_SEGMENT_SIZE;
+        }
+
+        protected AnnotatedTextFieldType(AnnotatedTextFieldType ref) {
+            super(ref);
+
+            this.fielddata = ref.fielddata;
+            this.fielddataMinFrequency = ref.fielddataMinFrequency;
+            this.fielddataMaxFrequency = ref.fielddataMaxFrequency;
+            this.fielddataMinSegmentSize = ref.fielddataMinSegmentSize;
+        }
+        
+        @Override
+        public void setIndexAnalyzer(NamedAnalyzer delegate) {
+            if(delegate.analyzer() instanceof AnnotationAnalyzerWrapper){
+                // Already wrapped the Analyzer with an AnnotationAnalyzer
+                // We
+                super.setIndexAnalyzer(delegate);
+            } else {
+                // Wrap the analyzer with an AnnotationAnalyzer that will inject required annotations
+                super.setIndexAnalyzer(new NamedAnalyzer(delegate.name(), AnalyzerScope.INDEX,
+                    new AnnotationAnalyzerWrapper(delegate.analyzer())));
+            }
+        }
+
+        public AnnotatedTextFieldType clone() {
+            return new AnnotatedTextFieldType(this);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (super.equals(o) == false) {
+                return false;
+            }
+            AnnotatedTextFieldType that = (AnnotatedTextFieldType) o;
+            return fielddata == that.fielddata
+                    && fielddataMinFrequency == that.fielddataMinFrequency
+                    && fielddataMaxFrequency == that.fielddataMaxFrequency
+                    && fielddataMinSegmentSize == that.fielddataMinSegmentSize;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(super.hashCode(), fielddata,
+                    fielddataMinFrequency, fielddataMaxFrequency, fielddataMinSegmentSize);
+        }
+
+        public boolean fielddata() {
+            return fielddata;
+        }
+
+        public void setFielddata(boolean fielddata) {
```

Comment:
Should it be allowed to use fielddata ? This field can be used in the significant text aggregation and the annotated text should be indexed in a doc values field so maybe we can disallow fielddata entirely ?

Reply:
>maybe we can disallow fielddata entirely ?

I wouldn't be opposed to that - we're looking to phase it out so this makes sense.